### ISSUE TITLE
View Z-Far Distance Above 5000 Creating Lighting Issue

### ISSUE THREAD

**Issue by joshchristiane** [godotengine/godot#42390]
Godot Version 3.2.2 Stable
Windows 10 Pro, AMD Ryzen, RTX 2070

Description:
When you increase the view distance of the main viewport (view z-far distance setting), objects close-up appear to have lines all over them in a weird pattern, see image attachment for example. This is an issue in every project I've tested so far.

![editor_screenshot_2020-09-28T135700-0800](https://user-images.githubusercontent.com/70781564/94485930-86e66b00-0193-11eb-81b0-91a3d5dda6c0.png)

**Comment by Calinou** [comment]
Duplicate of https://github.com/godotengine/godot/issues/13129. This was fixed in the `master` branch but not the 3.2 branch.

PS: You generally don't want to set the Z far distance so high. See discussion in https://github.com/godotengine/godot/pull/33207.

**Comment by clayjohn** [comment]
I tested this out and can confirm that it becomes a problem after a z-far of about 5000. However, I'm not sure what the solution is, or if it is worth finding one. I quickly tried using 32 bit depth buffers (for the scene and for the SSAO effect) and the issue didnt go away. This leads me to thinking it may actually be a precision issue in the shader. Which would require a much more complex solution. 

The only thing I can think of that would solve this in a satisfactory way would be to cap out the depth of the SSAO effect to something more reasonable like 1000 meters. Anything beyond that just wouldn't receive SSAO. However, even this solution would introduce significant complexity that may not be worth it.

**Comment by joshchristiane** [comment]
I agree that generically one shouldn't use a z-far beyond 2000 or maybe 3000... But it is still a problem that should be addressed at some point, especially for games with really big landscapes or calls from large objects really far away. Perhaps a solution could be a simple bool tick box added to objects to allow them to exempt from the z-far "filter" or whatever it's technically called, thus they would always render and be unaffected by the z-far distance. Perhaps there is a way to write a script to do this that I'm unaware of, but I could definitely find a use for that if I can't go beyond say 5000 meters.

**Comment by Calinou** [comment]
> I agree that generically one shouldn't use a z-far beyond 2000 or maybe 3000... But it is still a problem that should be addressed at some point, especially for games with really big landscapes or calls from large objects really far away. Perhaps a solution could be a simple bool tick box added to objects to allow them to exempt from the z-far "filter" or whatever it's technically called, thus they would always render and be unaffected by the z-far distance. Perhaps there is a way to write a script to do this that I'm unaware of, but I could definitely find a use for that if I can't go beyond say 5000 meters.

You can probably do something like this in Godot using a shader: https://aras-p.info/blog/2019/02/01/Infinite-sky-shader-for-Unity/
